fix Non-empty orders in the cart state do not appear in orders page

abdellani opened the pull request:
#### What? Why?

- Closes #11120

<!-- Explain why this change is needed and the solution you propose.
     Provide context for others to understand it. -->
I edited the SearchOrders to include the orders that are without a billing address.


#### What should we test?
<!-- List which features should be tested and how.
     This can be similar to the Steps to Reproduce in the issue.
     Also think of other parts of the app which could be affected
     by your change. -->

- described on: https://github.com/openfoodfoundation/openfoodnetwork/issues/11120#issue-1775250227
- 

#### Release notes

<!-- Please select one for your PR and delete the other. -->

Changelog Category: Technical changes

<!-- Choose a pull request title above which explains your change to a
     a user of the Open Food Network app. -->

The title of the pull request will be included in the release notes.


#### Dependencies
<!-- Does this PR depend on another one?
     Add the link or remove this section. -->



#### Documentation updates
<!-- Are there any wiki pages that need updating after merging this PR?
     List them here or remove this section. -->

abdellani commented on app/services/search_orders.rb:
```diff
@@ -19,7 +19,7 @@ def fetch_orders
       ransack(params[:q])
 
-    @search.result(distinct: true).joins(:bill_address)
+    @search.result.left_joins(:bill_address).uniq
```
When I tried the following query, The query fails when it comes to sort the order by the billing address 
check `spec/controllers/api/v0/orders_controller_spec.rb:143`
```ruby
@search.result(distinct: true).left_joins(:bill_address)
```
and with this
```ruby
@search.result.left_joins(:bill_address).uniq
```
pagy fails on  `Api::V0::OrdersController` and multiple tests fail.
```ruby
...
        if pagination_required?
          @pagy, orders = pagy(orders,
                               items: params[:per_page] || default_per_page)
        end
...
```

Matt-Yorkley commented:
Hmmm is it related to whether or not a shipping method has been selected (which happens during the first checkout step)? It looks like the `SearchOrders` service is being used here and it filters out orders without a shipping method?

https://github.com/openfoodfoundation/openfoodnetwork/blob/fe6075319e4482c80ecd7e2d62a6e17cc270ae43/app/services/search_orders.rb#L32-L36

abdellani commented:
> Hmmm is it related to whether or not a shipping method has been selected (which happens during the first checkout step)? It looks like the `SearchOrders` service is being used here and it filters out orders without a shipping method?


The [described issue](https://github.com/openfoodfoundation/openfoodnetwork/issues/11120) comes from the inner join between the query that loads the order and the bill address tables.

https://github.com/openfoodfoundation/openfoodnetwork/blob/99daab07b757e848917d3070ea473fa86f7a7fe1/app/services/search_orders.rb#L20-L23

Every order that's  
1. not empty
2. has no billing address
will not appear on the admin dashboard.

abdellani commented:
@Matt-Yorkley 
When I added that line of code, other tests failed. So, I reverted it.


The only test that's failing at this level is `./spec/controllers/api/v0/orders_controller_spec.rb:143`

```ruby
        it 'can sort orders by bill_address.lastname' do
          get :index, params: { q: { completed_at_not_null: true,
                                     s: 'bill_address_lastname ASC' } },
                      as: :json

          expect(json_response['orders']
            .map{ |o| o[:id] }).to eq serialized_orders([order2, order3, order1, order4])
              .map{ |o| o["id"] }
        end
```
I played with the UI and noticed that the test is not trying a parameter available on the UI. 
![image](https://github.com/openfoodfoundation/openfoodnetwork/assets/16113997/927ed459-d569-4152-a0f7-00ce86842885)

We're using `billing_address_name` instead of `bill_address_lastname`.

mkllnk reviewed (approved):
Nothing is easy, eh? :wink: The result of all your searching looks really good though.

dacook commented on app/services/search_orders.rb:
```diff
@@ -19,7 +19,7 @@ def fetch_orders
       includes(:payments, :subscription, :shipments, :bill_address, :distributor, :order_cycle).
       ransack(params[:q])
 
-    @search.result(distinct: true).joins(:bill_address)
+    @search.result(distinct: true).left_joins(:bill_address)
```
I was thinking that this join statement should be included earlier, then noticed that that the association is already mentioned in the `includes` statement. Wouldn't this have the effect of a left join, and so we can simply remove the  `joins(:bill_address)` part altogether?

I couldn't easily confirm this myself, because this bug wasn't covered by a spec.

dacook replied:
I decided to go ahead and inspect the SQL, current, and after removing `.left_joins(:bill_address)`( because bill_address is already included in `includes`):

```diff
 SELECT DISTINCT spree_addresses.*,
                 spree_orders.*
 FROM   spree_orders
        LEFT OUTER JOIN spree_line_items
                     ON spree_line_items.order_id = spree_orders.id
        LEFT OUTER JOIN spree_addresses
                     ON spree_addresses.id = spree_orders.bill_address_id
-       LEFT OUTER JOIN spree_addresses bill_addresses_spree_orders
-                    ON bill_addresses_spree_orders.id = spree_orders.bill_address_id
 WHERE  ( spree_orders.distributor_id IN (SELECT enterprises.id
                                          FROM   enterprises)
           OR spree_orders.order_cycle_id IN (SELECT order_cycles.id
                                              FROM   order_cycles) )
        AND ( spree_line_items.id IS NOT NULL
               OR spree_orders.state IN ( 'complete', 'canceled', 'resumed',
                                          'awaiting_return', 'returned' ) )
 ORDER  BY spree_addresses.lastname ASC
```
(I used an online code formatter for legibility too)

We can see that there was a duplicate LEFT OUTER JOIN on `spree_addresses`, which doesn't seem to be referred to at all. I tested this with the new specs and it continues to pass.

So I think we can remove it, do others agree?

dacook replied:
Actually I have to admit I don't understand why `bill_adresses` is included with a left join, while other tables in the `includes` aren't.
In any case, the tests seem to indicate that we don't need the additional bill_addresses join, so I've pushed a suggested commit to demonstrate. Feel free to remove again if it's inappropriate.

mkllnk replied:
> Actually I have to admit I don't understand why bill_adresses is included with a left join, while other tables in the includes aren't.

The Rails `includes` method can use either a left join or a separate query:

> With includes, Active Record ensures that all of the specified associations are loaded using the minimum possible number of queries.
> https://guides.rubyonrails.org/active_record_querying.html#eager-loading-associations

There are two ways to tell Rails more specifically what you want:

* `references`: Tell Rails to use a left join because you have `select` or `where` statements referencing that table.
* `eager_load`: Instead of `includes`, this one always does a left join. It's like `includes(:table).references(:table)`.
* `preload`: Instead of `includes`, this one always loads in a separate query.


So Rails has to basic mechanisms: eager_load and preload. If you don't care which one is used and you want Rails to decide that when the full query is executed then you can use includes and it will use one of them under the hood. If you want to automatic `includes` in some cases but in some you want to reference the the table then you should declare that.

For this particular case of bill_address, Rails seems to do the right thing already. I'm wondering if Ransack is clever enough to tell Rails that it's referencing that association. Or maybe we do something already that makes Rails think that we need that join. Since we do a `select` we should probably add the `references` to be sure Rails joins. But since there were two joins before, Rails may see two different associations. We need to check that `references` isn't adding one again. Is it maybe that we use `bill_address` in some cases and `billing_address` in others? I noticed somewhere that we using *billing* in a Ransack query because there were issues with *bill* but maybe with this knowledge there's a better way to resolve it.

dacook replied:
Wow, it's complex. Thanks for diving in, but I'm not feeling super clear if it's worth changing anything?

It appears we have it covered by specs, and they are passing so I'd avoid adding new code.
But... I just tried adding a spec for _billing_ address, and it broke due to my change. Ok I will revert my change!

But... the order is wrong for _billing_ address, so we don't seem to support a query on this alias. Do we need to?

dacook reviewed (commented):
Good find. I guess it turned out this bug was unrelated to https://github.com/openfoodfoundation/openfoodnetwork/pull/8148

I wondered if we can remove the join statement altogether, what do you think?

abdellani commented:
> Good find. I guess it turned out this bug was unrelated to #8148
> 
> I wondered if we can remove the join statement altogether, what do you think?

I don't know to be honest. It looks like it's needed for sorting the order.
This join was included on this [commit](https://github.com/openfoodfoundation/openfoodnetwork/pull/8615/commits/35e20fc179da821ca2a65f19b30ff90b79e90621)

Maybe @Matt-Yorkley  or @jibees  can help us with that?

Matt-Yorkley commented:
Hmmm so we're changing the `SearchOrders` service here and the thing to note with regards to testing is that it gets used by both the regular `/admin/orders` page and by the bulk order management page as well (via `Admin::OrdersController` and `Api::V0::OrdersController` respectively). The search filters that get passed to the service are slightly different in both cases (it's not possible to search/filter by name on the BOM page).

Matt-Yorkley commented:
I think we need to add a bit more test coverage here for the specific scenario where there's an order without a billing address and we're also using the billing address name for filtering. Example here: https://github.com/openfoodfoundation/openfoodnetwork/commit/ccf436dcd8fb35f04744646f21ac118dd55a16f4

The second of those two tests fails for me (with the changes in this PR) even after updating the search scope from `bill_address_lastname` to `billing_address_name`. That change to the way the scopes work for ordering by first and last names (combined) was added here FYI: https://github.com/openfoodfoundation/openfoodnetwork/pull/11299 (but the related tests for this controller were not updated at the time, as you spotted).

abdellani commented:
thank you @Matt-Yorkley :pray: 
I added your tests.
That was really helpful. I fixed the issues without a need to edit the existing test now.

abdellani commented:
> Good find. I guess it turned out this bug was unrelated to #8148
> 
> I wondered if we can remove the join statement altogether, what do you think?

I think that the join is only necessary when we need to sort the orders by `billing_address.XYZ`  :+1:

dacook reviewed (commented):
I've added a commit to reduce the code as per [my comment](https://github.com/openfoodfoundation/openfoodnetwork/pull/11348#discussion_r1303648697). @abdellani does this look ok to you?

mkllnk commented on spec/controllers/api/v0/orders_controller_spec.rb:
```diff
@@ -175,6 +175,14 @@ module Api
           end
+
+          it 'can sort orders by billing_address.lastname' do
+            get :index, params: { q: { s: 'billing_address_lastname ASC' } },
```
I can't find this parameter in the codebase, only `billing_address_name`. Shall we use a real example? Interestingly the scope for ransack, `sort_by_billing_address_name_asc`, is doing a `join` and therefore ignoring records without address. I'm a bit lost in this mess. Do you have time to look into this, @dacook?

dacook replied:
Honestly I'm a bit lost too, I just came in to review this little chunk of code.. 
I'm hoping that @Matt-Yorkley will have prior knowledge, and be able to to say what's what?

mkllnk replied:
I feel confident that I could work this out but I would need to spend a bit of time on it. Maybe tomorrow unless Matt looked at this already.

mkllnk commented on app/services/search_orders.rb:
```diff
@@ -15,11 +15,19 @@ def initialize(params, current_user)
   def fetch_orders
-    @search = search_query.
+    # Remap "billing" address
+    params.dig(:q, :s)&.gsub!("billing_address", "bill_address")
```
I just checked out this branch and navigated to `/admin/orders` and got an error:
```
FrozenError in Spree::Admin::OrdersController#index 
can't modify frozen String: "completed_at desc"
```

Looks like we are missing some test coverage here!

mkllnk replied:
> Looks like we are missing some test coverage here!

Actually not, lots of specs are failing. I just didn't pay attention because we have so many flaky specs and I assumed that @dacook would have checked the spec outcome. :see_no_evil:

mkllnk replied:
This change also broke the sorting function because we have custom search scopes `sort_by_billing_address_name_[asc,desc]` and changing that to `bill` breaks it. Sorry, @dacook, I'll take your commit out again.

dacook replied:
Sorry, yes it was a bit of an experiment on my part 😬 .

mkllnk commented:
Uh, just discovered that Filipe wrote a pending spec for this issue which we should activate in this PR. It was mentioned in the issue but apparently forgotten:

* https://github.com/openfoodfoundation/openfoodnetwork/pull/11121

mkllnk commented on app/models/spree/order.rb:
```diff
@@ -149,11 +149,13 @@ def states
     }
 
     scope :sort_by_billing_address_name_asc, -> {
-      joins(:bill_address).order("spree_addresses.lastname ASC, spree_addresses.firstname ASC")
+      references(:bill_address)
+        .order("spree_addresses.lastname ASC, spree_addresses.firstname ASC")
```
This happens when sorting on the normal orders page. It's a custom search scope and Ransack doesn't add any magic. The original query has `includes(:bill_address)` already which we could repeat here but I kept it minimal and just used `references`.

dacook replied:
I think it would be worth a code comment here explaining why the `references` needs to be here. (Of course, the specs will be the ultimate protection).

mkllnk commented on app/services/search_orders.rb:
```diff
@@ -15,11 +15,16 @@ def initialize(params, current_user)
+      result(distinct: true)
 
-    @search.result(distinct: true).joins(:bill_address)
+    if params.dig(:q, :s)&.starts_with?("bill_address_")
+      search = search.select('spree_addresses.*, spree_orders.*')
+    end
```
This parameter comes from the bulk order management via the API controller it's using Ransack features to find the attribute on the association. Ransack seems to add a join for this and we would end up with two joins if we added `references` here as well. So we don't and just need to help Ransack with the select.

mkllnk commented on spec/system/admin/orders_spec.rb:
```diff
@@ -269,17 +269,20 @@
       end
 
       it "displays non-empty cart orders" do
-        pending "issue #11120"
-
```
Thanks to Filipe, this spec was already here. :smile:

mkllnk commented on spec/system/admin/orders_spec.rb:
```diff
@@ -269,17 +269,20 @@
 
-        # pending issue #11120
+        # And the same orders are displayed when sorting by name:
+        find("th a", text: "NAME").click
+
+        expect(page).to have_no_content order_empty.number
+        expect(page).to have_content order_not_empty.number
```
This extends the spec. I kept it in the same block to save on execution time. There was another inner join in this case which was hiding orders without bill address.

mkllnk reviewed (approved):
Two hours later I finally think that I know what's happening here. It may be better to use the same search scope for the bulk order edit page and then we can reduce the code but it works for now. One day we will replace AngularJS...

dacook commented on spec/system/admin/orders_spec.rb:
```diff
@@ -277,6 +277,13 @@
+        # And the same orders are displayed when sorting by name:
+        find("th a", text: "NAME").click
+
+        expect(page).to have_no_content order_empty.number
```
I thought that Capybara actually treated this the same way in more recent versions:

```suggestion
        expect(page).not_to have_content order_empty.number
```

But I can't immediately find proof of this. I think Rubocop should tell us which one to use..

mkllnk replied:
Yes, it's treated the same way. Do you have a preference? Reasons for or against?

dacook replied:
I think if there's no cause for a special matcher, then we should use the one way. It's slightly easier to see the `not` when it's in the expected position.

Of course, the difference is ever so slight.. 
I'll add it to my growing list of rules to suggest.

dacook reviewed (approved):
Great work!
It looks great from my point of view.
